### PR TITLE
Add support for special placeholder files in build_test.

### PR DESCRIPTION
**Pull request by matanlurey**
Closes https://github.com/dart-lang/build/issues/782.

### CODE REVIEW

**Comment by matanlurey** [comment]
@jakemac53:

> Technically attempting to read these assets is supposed to throw, and they shouldnt show up in globs either, but I don't know how important that is to support fully.

I'm happy to add that to at least the build_test package before committing, though it does mean changing the `expectedOutputs` function, which I'm not sure what the impact would be.

File: build_test/lib/src/test_builder.dart [hunk 1]
```diff
@@ -119,6 +119,15 @@ Future testBuilder(
     inputIds.add(id);
   });
 
+  var allPackages = reader.assets.keys.map((a) => a.package).toSet();
+  for (var pkg in allPackages) {
+    for (var dir in const ['lib', 'web', 'test']) {
+      var asset = new AssetId(pkg, '$dir/\$$dir\$');
+      reader.cacheBytesAsset(asset, const []);
+      inputIds.add(asset);
```
Comment (jakemac53) [inline build_test/lib/src/test_builder.dart]: What if we only do this and just remove the line above that adds the asset to the reader? I think that might give the desired behavior.

Comment (matanlurey) [reply build_test/lib/src/test_builder.dart]: Sure. I'll tackle this before merging.

Comment (matanlurey) [reply build_test/lib/src/test_builder.dart]: Done.

**Comment by jakemac53** [comment]
I was just thinking about this and I think it is actually a breaking change, at least for build_runner?

That isn't a huge issue, but we should release it as such.

**Comment by matanlurey** [comment]
@jakemac53:

> I think it is actually a breaking change, at least for build_runner

Can clarify what you mean here? You mean we need a newer version of BR as a lower constraint?

**Comment by jakemac53** [comment]
> Can clarify what you mean here? You mean we need a newer version of BR as a lower constraint?

What I meant is this is a breaking change in the build_test package in general because there are now additional inputs that will be matched. If you were expecting certain outputs before and had an overzealous builder (like CopyBuilder), you will either have new outputs, or get runtime errors when you try to read these files. Both would cause your tests to fail.

**Comment by matanlurey** [comment]
> What I meant is this is a breaking change in the build_test package

Oh ok. I agree. Do we want to ship `0.9.3` as-is? I can put this behind a flag and not enable yet.

Or we can skip to `0.10.0`.

/cc @natebosch

**Comment by jakemac53** [comment]
I don't think we need to ship the current version, we can just delete `CopyBuilder` and release that with this change.

**Comment by matanlurey** [comment]
PTAL @jakemac53 at the CHANGELOG before I merge.

**Pull request merged**